latch_cli/centromere/ctx.py

```diff
@@ -0,0 +1,264 @@
+
+                self.dkr_client = construct_dkr_client(
+                    ssh_host=f"ssh://{username}@{public_ip}"
+                )
+                self.ssh_client = construct_ssh_client(
+                    host_ip=public_ip,
+                    username=username,
+                )
+            else:
+                self.dkr_client = construct_dkr_client()
+        except Exception as e:
+            self.cleanup()
+            raise e
+
+    @property
+    def image(self):
+        """The image to be registered."""
+        if self.account_id is None:
+            raise ValueError("You need to log in before you can register a workflow.")
+
+        # CAUTION ~ this weird formatting is maintained indepedently in the
+        # nucleus endpoint and here.
+        # Name for federated token request has minimum of 2 characters.
+        if int(self.account_id) < 10:
+            account_id = f"x{self.account_id}"
+        else:
+            account_id = self.account_id
+
+        return f"{account_id}_{self.pkg_root.name}"
+
+    @property
+    def image_tagged(self):
+        """The tagged image to be registered.
+
+        eg. dkr.ecr.us-west-2.amazonaws.com/pkg_name:version
+        """
+        if self.version is None:
+            raise ValueError(
+                "Attempting to create a tagged image name without first "
+                "extracting the package version."
+            )
+
+        # From AWS:
+        #   A tag name must be valid ASCII and may contain lowercase and uppercase letters,
+        #   digits, underscores, periods and dashes. A tag name may not start with a period
+        #   or a dash and may contain a maximum of 128 characters.
+
+        match = re.match("^[a-zA-Z0-9_][a-zA-Z0-9._-]{,127}", self.version)
+        if not match or match.span()[0] != 0 or match.span()[1] != len(self.version):
+            raise ValueError(
+                f"{self.version} is an invalid version for AWS "
+                "ECR. Please provide a version that accomodates the ",
+                "tag restrictions listed here - ",
+                "https://docs.aws.amazon.com/AmazonECR/latest/userguide/ecr-using-tags.html",
+            )
+
+        if self.image is None or self.version is None:
+            raise ValueError(
+                "Attempting to create a tagged image name without first "
+                " logging in or extracting the package version."
+            )
+        return f"{self.image}:{self.version}"
+
+    def task_image_name(self, task_name: str) -> str:
+        return f"{self.image}:{task_name}-{self.version}"
+
+    @property
+    def full_image(self):
+        """The full image to be registered (without a tag).
+
+            <repo/image>
+
+
+        An example: ::
+
+            dkr.ecr.us-west-2.amazonaws.com/pkg_name
+
+        """
+        return f"{self.dkr_repo}/{self.image}"
+
+    def nucleus_provision_url(self) -> (str, str):
```

Comment:
instead of making this a method, have the ip, username be class fields, and hit the endpoint in the __enter__ method (that way we don't have to explicitly call this)

Reply:
i really dont see what that would do, we would have to explicitly call either way (one in __init__ and other in __enter__) with the way it is now, just feed the values directly into the docker + ssh clients rather than creating unnecessary class attributes

Reply:
nm changed